SSA/Core/MLIRSyntax/Parser.lean

```diff
@@ -0,0 +1,132 @@
+
+unsafe def elabIntoMeta {α : Type} (expectedType : Expr) (stx : Lean.Syntax) : MetaM α :=
+  elabIntoTermElab (α := α) expectedType stx |>.run'
+
+unsafe def elabIntoCore {α : Type} (expectedType : Expr) (stx : Lean.Syntax) : CoreM α :=
+  elabIntoMeta (α := α) expectedType stx |>.run'
+
+def printException : Except Exception α → IO String
+  | Except.ok _ => throw <| IO.userError "printException called on Except.ok"
+  | Except.error e => e.toMessageData.toString
+
+unsafe def elabIntoEIO {α : Type} (env : Lean.Environment) (typeName : Lean.Expr) (stx : Lean.Syntax) : EIO ParseError α :=
+  fun s =>
+    let resE : EIO Exception α := elabIntoCore (α := α) typeName stx |>.run' {fileName := "parserHack", fileMap := default} {env := env}
+    match resE s with
+    | .ok a s => .ok a s
+    | .error exception s =>
+      let errMsgIO : IO String := exception.toMessageData.toString
+      match errMsgIO s with
+      | .ok errMsg s => .error s!"failed elaboration {stx}. Error: {errMsg}" s
+      | .error e s => .error s!" Failed elaborating {stx}.\nUnable to pretty-print exception at 'elabIntoEIO':\n{e}." s
+
+def region0Expr := (Expr.app (Expr.const `MLIR.AST.Region []) (Expr.const `Nat.zero []))
+def op0Expr := (Expr.app (Expr.const `MLIR.AST.Op []) (Expr.const `Nat.zero []))
+/--
+  Parse `Lean.Syntax` into `MLIR.AST.Region`.
+
+  This runs the Lean frontend stack and adds it to the TCB.
+  It should be safe (if we trust the lean parser) since we ensure that the types match: see
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabRegion (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Region 0) := do
+  let reg ← unsafe elabIntoEIO (α := MLIR.AST.Region 0) env region0Expr stx
+  return reg
+
+/--
+  Parse `Lean.Syntax` into `MLIR.AST.Op`.
+
+  This runs the Lean frontend stack and adds it to the TCB.
+  It should be safe (if we trust the lean parser) since we ensure that the types match: see
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabOp (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Op 0) :=
+  unsafe elabIntoEIO (α := MLIR.AST.Op 0) env op0Expr stx
+
+def mkParseFun (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (s : String) (env : Environment) : EIO ParseError ParseOutput := do
+  match Parser.runParserCategory env syntaxcat s with
+    | .error msg => throw msg
+    | .ok syn => ntparser syn
+
+private def mkNonTerminalParser (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (env : Environment) (s : String) : EIO ParseError ParseOutput :=
+  let parseFun := mkParseFun syntaxcat ntparser
+  parseFun s env
+
+-- TODO: do we need this copy of the environment, or can I remove it from one of the two?
```

Comment:
Should this be resolved, still?

Reply:
are you asking if this TODO is still necessary or if we can address it and then remove it after having addressed it? I don't know how to do the latter, but if you want to take a stab at it, that'd be very helpful!

Reply:
@goens I'm not sure what the comment means --- I'm trying to address the TODO, but it's unclear to me what I would need to change to test this. Can I have a sentence or two explaining what environment copy I can try and remove?

Reply:
Just an actionable comment would be enough. Just a bit of information that would people like @bollu enable to pick this issue in the future (or now) would be helpful.